### DISCUSSION TITLE
ElectroDacus SBMS0

### DISCUSSION THREAD

**Discussion by Louisvdw** [Louisvdw/dbus-serialbattery#556] [New BMS requests]
- [ ] SOC
- [ ] Battery voltage
- [ ] Charge/discharge current
- [ ] Temprature
- [ ] Min/max cell voltages
- [ ] protection notifications

**Comment by willobrien** [comment]
+1 on this. IMO, what's needed is a gateway to ship BMS values across canbus.
On my camper build, just some form of integration would be great. On my boat, everything must speak canbus: BMS, Cerbo GX, WS500 alternator controller...  I'm using a different BMS on the boat, but am getting tempted to install a SBMS0 for monitoring/exposing the data a bit more. Whatever comes from this, will be useful to me. Happy to help.

**Comment by monophono** [comment]
Any news about the SBMS integration?

**Comment by charlesrg** [comment]
Would love to heart about this, anything that I could help ?

**Comment by charlesrg** [comment]
I guess the workaround will be CerboGX with NodeRed and MQTT on the SBMS0 Wifi

**Comment by diegotam855** [comment]
I connected the SBMS0 to Venus OS in NodeRed with MQTT but It would be awesome to directly connect the BMS to Venus and see the cells voltages in the VRM

**Reply by CaptainRivet** [reply]
Could you please share the integration.
Maybe a little how to would be great as i have no idea about NodeRed and Mqqt...just starting and need to start somewhere.

**Reply by diegotam855** [reply]
First you need to have installed the large version of VenusOS:
![1](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/28c6790b-3e1a-4634-b12a-76cd9725d47b)
![2](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/9eb4c027-6e36-4ceb-a0fb-587b358594f5)

You now have a new menu:
![4](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/3aa6ea35-e923-483f-afb0-03d6f3ad5186)

Where you need to enable NodeRed:
![5](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/8758db8e-9163-4aa0-9a84-9fb086f21bb0)
You can access NodeRed on the port 1881 or remotely via VRM portal. 

Than MQTT needs to be enabled:
![3](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/c76925ac-3812-41c5-a0ec-729ac62dc9f0)

**Reply by diegotam855** [reply]
In the SBMS0 setting page you need to insert your MQTT broker IP and port. 
I'm using a Teltonika RUT240 router as a broker, it has this function integrated. So my IP is the IP of the router.
After this the BMS is sending data to the broker (in my case the router) and now in NodeRed you can subscribe to this data stream and use it as you want.
Of course the BMS needs to be connected via wifi to the same network.

![6](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/12a95c46-3121-4223-893e-e6a1e20fb3cf)

**Reply by CaptainRivet** [reply]
Thank you, really appreciated.
Running gui mods so large is already installed and rest on cerbo just done due to your description. 
Whats that broker stuff on the router? I have a GL-inet AL750 slate. The SBMS is connected to router via fixed ip 192.168.8.11. What do i need to do on the router?
Then install the serial-bat on the cerbo, like its described in the read me, no chances or adaption?
How reliable is the wifi? 
I am using the web page interface and that freezes minimum once a week, disable and then enable wifi on SBMS fixes it in 95%, rest 5% power cycle router.

**Reply by CaptainRivet** [reply]
Then install serial bat on cerbo, no adaption or changes?

**Reply by diegotam855** [reply]
Gui Mods is not related with the large installation of VenusOS they are two independent things as I know...I may be wrong though.
You need to understand what MQTT is and the basics on how it works. I did the same watching a dozen YouTube videos. I suggest this one as a start: [https://www.youtube.com/watch?v=NjKK5ab0-Kk](url)
Btw I'm using a RaspberryPi not the Cerbo.

Basically you need one device in your network that can host a software that act like a server for the MQTT protocol. I used the router but you can use any computer. 

I didn't had any problem with the WIFI connection, it never failed once.

**Reply by mr-manuel** [reply]
Venus OS has an integrated MQTT Broker

**Reply by diegotam855** [reply]
Oh really!? I didn't know that I can use that, how do I get access to it?

**Reply by mr-manuel** [reply]
Just enable it under `Settings` -> `Services` -> `MQTT on LAN (SSL)` and `MQTT on LAN (Plaintext)`.

SSL is on port `tcp/8883`
Plaintext is on port `tcp/1883`

**Reply by CaptainRivet** [reply]
So that means i can simply put into MQTT connection of electrodacus BMS under server ip the ip of the cerbo and under server port the 8883. Or do i need to use plaintext?

SSL is on port tcp/8883
Plaintext is on port tcp/1883

**Reply by mr-manuel** [reply]
Yes. If SSL do not work use plaintext, if the connection is in a secure network.

**Reply by CaptainRivet** [reply]
First i change/set all according to diegotam's description 
2nd setting server in ElectrodacusBMS to server ip of cerbo (in my case 168.192.8.138) and port 1883
3rd : install dbus serial due to this tutorial via SSH https://louisvdw.github.io/dbus-serialbattery/general/install/
Then all should work, correct?

**Reply by CaptainRivet** [reply]
@diegotam855 So got all adapted due to your description and installed the dbus-serialbattery via SSH on cerbo. Now the last piece Node red is missing. Could you please share the NodeRed flow. Need to find the service i can subscribe to....Tried dinding it connecting via MQTT explorer to the SBMS but no luck, doesn't connect via IP and port 1883. Cerbo works via MQTT explorer but i cannot find anywhere the ElectrodacusBMS. Can you help please.

**Reply by diegotam855** [reply]
Try to import my flow and change the MQTT node server IP accordingly to your network.
Mine is the router as I said before. 
I don’t know how it works when the server is on the same machine as Node Red.. 

Ignore the bottom part, that’s to control a  relay to turn on an active cell balancer based on cell voltages. 

[flows.json](https://github.com/Louisvdw/dbus-serialbattery/files/13814195/flows.json)

**Reply by CaptainRivet** [reply]
Great, thank you. I like to turn on an active balancer too, so thats welcome :-)
Done all but the Battery monitor doesn't show up. Do i need to change something in serialbattery like config file? 
How can i check/debug the different steps. 1) ElectrodacusBMS  via Node red to cerbo 2) cerbo/venusOS serialbattery.
With root and dbus-spy i don't see the d-bus-serialbattery.

**Reply by CaptainRivet** [reply]
In the NodeRed flow i see "SBMS" receive node connected. I assume that means node red gets mqtt data from SBMS

**Reply by CaptainRivet** [reply]
Which BMS of dbus-serialbattery is used by tge cerbo?

**Reply by CaptainRivet** [reply]
@diegotam855 could you please help again, i am somehow stuck. thank you

**Reply by diegotam855** [reply]
I never own a Cerbo so I really can't help with that. And I'm not an IT expert, I learned like you by trial and error. 

As far as I know if the MQTT node in Node Red has a green dot says connected it just means you have connection to the broker. Than you need to be subscribed to the correct topic. If in the settings of that node under "topic" you have **"/sbms"**  and in the settings page of SBMS0 under "MQTT prefix" you have just a slash **"/"** it should work. 
Assuming the devices are connected to the same network and IP addresses and ports are set correctly.
Unfortunately there is no way, as far as I know, to see if the SBMS0 is connected to the broker and it's sending data.

Sorry if I said obvious things but just to be clear..

**Reply by CaptainRivet** [reply]
@diegotam855 thanks for the answer. Cerbo or raspi are basically the same. both have the same Venus OS installed and therfor also broker. only difference raspi has one more usb but has no digital and anaolg ports the cerbo has. another difference is the video interface, not iinvolved here. so for the dbus-serialbattery no difference between Raspi or cerbo.
Ok understand we are in the same boat/IT level, the difference you have it running and i have to figure how it must work.

thanks for stating the obvious again, i have it exactly like this.
Looks like node-Red and MQTT is working like it should. i get green dot, states connected and all seems to work here when i deploy no mistakes in debug.
A dumb question: I don't have to connect the Electrodacus with a USB cable to raspi/cerbo? assume not but how is it connected as the driver is USB or bluetooth?
i checked if the serialbus-driver is working and I think there the problem is lying:
under which BMS manufacturer the Electrodacus connects in your setup means eg JK-BMS or Seplos or Heltek or JBD....there is no Electrodacus so it must fit with another BMS manaufacturer but which?
Should show this in the devicelist of the remote console.
andhow is it connected to then  USB0-USB4 or does it use the BT serialbus connection?
going through troubleshooting here: https://louisvdw.github.io/dbus-serialbattery/troubleshoot/#datalogserial-startercurrent

@mr-manuel : my you could help too, will make then a tutoial for the Electrodacus community...i am close i know...
i am using venus OS 3.13 and bus-serialbattery v1.0.20230531 driver, installed via SSH and using 1 latest build.
reinstalled driver and the same...

to see this could you diegotam do me the favour and in putty with remote access run the following:
tail -F /data/log/serial-starter/current | tai64nlocal
you can stop it with CTRL+C
it should show you the list where its connected and on mine its not connecting
2024-01-07 18:12:10.536781500 INFO: Start service vedirect-interface.ttyUSB1 once
2024-01-07 18:12:11.715353500 INFO: Start service vedirect-interface.ttyUSB2 once
2024-01-07 18:12:15.000756500 INFO: Start service dbus-serialbattery.ttyUSB1 once
2024-01-07 18:12:16.359028500 INFO: Start service dbus-serialbattery.ttyUSB2 once
2024-01-07 18:12:17.948927500 INFO: Start service gps-dbus.ttyUSB4 once
so it starts the service which is the good news but cannot connect to any USB port which is the bad news. just tries all 4 but doesn't connect.
also this should show which port und which BMS is connected.
tail -F -n 100 /data/log/serial-starter/current | grep dbus-serialbattery | tai64nlocal
here i get nothing.
but under 
head /data/var/lib/serial-starter/*
i get its connected: to USB0
==> /data/var/lib/serial-starter/ttyUSB0 <==
sbattery
to check all devices the serialstarter started:
tail -F /data/log/dbus-serialbattery.*/current | tai64nlocal
that looks like that...no BMS dedected at USB0, looks like he doesn't find the right manufacturer.
==> /data/log/dbus-serialbattery.ttyUSB0/current <==
2024-01-07 19:37:35.076868500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:35.084040500 INFO:SerialBattery:Testing LltJbd
2024-01-07 19:37:35.362202500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:35.371713500 INFO:SerialBattery:Testing Renogy
2024-01-07 19:37:35.661882500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:35.668931500 INFO:SerialBattery:Testing Renogy
2024-01-07 19:37:35.940773500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:35.948513500 INFO:SerialBattery:Testing Seplos

==> /data/log/dbus-serialbattery.ttyUSB1/current <==
2024-01-07 19:37:36.796404500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:36.799594500 INFO:SerialBattery:Testing HLPdataBMS4S

==> /data/log/dbus-serialbattery.ttyUSB0/current <==
2024-01-07 19:37:36.961078500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-01-07 19:37:37.462885500 ERROR:SerialBattery:ERROR >>> No battery connection at /dev/ttyUSB0
2024-01-07 19:37:37.463147500

then i check the runtime:
root@einstein:~# svstat /service/dbus-serialbattery.tty*
/service/dbus-serialbattery.ttyUSB0: up (pid 15372) 20 seconds
/service/dbus-serialbattery.ttyUSB1: down 12 seconds, normally up
/service/dbus-serialbattery.ttyUSB2: down 11 seconds, normally up
/service/dbus-serialbattery.ttyUSB4: up (pid 15249) 29 seconds

reinstalling driver I get the same.
so how to fix this?

**Reply by diegotam855** [reply]
@CaptainRivet I see some misunderstood base point here, I do not use serial battery connection to see the SBMS0 in the device list in the remote console. I never dealt with USB driver and such. I think is not even possible that's why I started to use NodeRed...
In my setup I receive data from the SBMS0 only in Node Red because I'm using MQTT. 
Second thing I took it for granted that your SBMS0 has Wi-Fi but maybe not? Mine is connected to my Wi-Fi network and it can send data to Node Red because of that.

As you can see here Wi-Fi client is enabled with the credential of my network.
![6](https://github.com/Louisvdw/dbus-serialbattery/assets/104109110/d2e63a0d-4b9f-43db-a24c-1c23f9a5afda)

In putty connected via SSH I can see all the VeDirect devices connected and nothing else.

**Reply by CaptainRivet** [reply]
Upsi, thats indeed a big missunderstanding. Just thought the serialbusdriver takes that data from the MQTT broker and displays it then as battery monitor in venus OS...thats obviously wrong. so gonna uninstall that.

Yes my Electrodacus has Wifi and i set up like you described.
then took your flow, changed the broker IP to the MQTT broker of venus OS which in my case is 192.168.8.138 with port 1883.
So the Electrodacus sends the data and Note red takes that jason streamas input node and your jason flow is then patching the data in the structure venus OS thinks its a battery monitor and displays it. Do I have to make any adjustments in Venus OS so it shows up as battery monitor?

**Reply by diegotam855** [reply]
Is far as I know that’s not possible. I’m using a smart shunt as battery monitor. 
Node Red is a completely separate software that works on its own. Even if it’s on the same machine. Then you can use the Victron nodes to read data and send commands to Venus. 
so if your sbms is connected to WiFi you should receive data in node red with my flow. 
Am I missing something?

**Reply by CaptainRivet** [reply]
does this data then create a direct VE battery monitor you can see via putty SSH in the device list?
i enabled MQTT and MQTT via LAN which is needed to switch on the MQTT server of Venus OS due to this description from victron:
https://community.victronenergy.com/questions/155407/mqtt-local-via-mqtt-broker.html
On the local broker in can connect with the MQTT explorer to local MQTT broker and see whats descibed by victron.
cannot see any data of SBMS in Venus OS, no battery monitor dedected but maybe because i connect via VRM and the local broker doesn't publish that to the cloud broker.
Sure it has something to do with MQTT broker. the local broker also gets a keep alive like described.
mmh maybe i try to connect to the cloud MQTT broker instead and see if that works

**Reply by mr-manuel** [reply]
The dbus-serialbattery will only work with a supported BMS. You cannot feed it with MQTT data.

If you want to add a battery with MQTT data you can use [dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery).

**Reply by diegotam855** [reply]
@CaptainRivet That's what you are looking for!  I already have installed the smart shunt so I'm not gonna rebuild my system. 
@mr-manuel is this showing up in the VRM portal as well?

**Reply by CaptainRivet** [reply]
I have the Victron battery monitor BMV712 installed on the bank too (missing the rj12 cable but on its way from china) which is equal to your smart shunt but has the advantage of a programmable relay that works in my system as last line of defense with HVC and LVC in case electrodacusBMS fails. It also has 2 measurement circuits means I use them for midpoint bank monitoring which checks cell 1+2 versus cell 3+4 means I have a big cell deviation the BMV712 can also discover the runner and cut the bank if the Electrodacus fails.
The issue with the battery monitor in Venus is it shows the voltage measured after the terminals only as total and also SOC is not what the BMS measures. But that’s not what the BMS shows and less precise plus only total voltage. It’s enough that DVCC working well but for monitoring via Venus os just not enough.

**Reply by CaptainRivet** [reply]
@diegotam855: your flow still doesn‘t Show me the values in node-Red, it’s the meat broker of Venus OS that’s not woking as it should. Figured I need to add keepalive so it continuously sends the data. Then I also need to somehow register the Matt client from Electrodacus so it passes the data over via the broker…found a video how to do that but no time execute that.

do I understand right that I then have the data only in node red but not in Venus OS. So to get it into Venus os I would need to add another flow connecting the values to Victron input nodes eg node Victron battery monitor.
I assume the data will only show up in Venus OS when the smart shunt or BMV712 in my case is connected as battery monitor and creates the instance for the battery monitor that will then display the additional data received by node red. Correct?

**Reply by mr-manuel** [reply]
> @mr-manuel is this showing up in the VRM portal as well?

@diegotam855 yes, otherwise it wouldn't make sense.
@CaptainRivet the [dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery) needs only his MQTT data for working properly. No other software/hardware required, but probably you need Node-RED to prepare the MQTT data for the driver.

A possible path of the data could be:

`SBMS0` -> MQTT Broker (Venus OS) with custom topic, e.g. `ElectroDacus/SBMS0/battery-1` -> Node-RED fetching the data from MQTT with the topic `ElectroDacus/SBMS0/battery-1`, prepare it for dbus-mqtt-battery and send it to MQTT with the topic `dbus-mqtt-battery/battery-1` -> The dbus-mqtt-battery then fetches the data from MQTT with the topic `dbus-mqtt-battery/battery-1` (need to be configured in the `config.ini`).

**Reply by CaptainRivet** [reply]
thank you both for being so patient with me and your help.
@diegotam855  : could you please provide the other flows as well that are linked to the provided jason file. if i understood right eg you need then a node that patches the MQTT data into victron node, assume the battery monitor (your connected smart shunt), correct?

@mr-manuel: gonna try the dbus-mqtt battery. Could you please help the mqtt broker in venus OS? Its all enabled in the remote console as per diegotams description. The node-Red tells me its connected but i get no data in the debug nodes . I assume my issue is from SBMS to MQTT broker of Venus. 
So the path here  he is SBMS0 -> local MQTT Broker (Venus OS) with custom topic /SBMS0
As you can see above SBMS screenshots from diego it has an MQTT client sending data with the topic /SBMS to the local MQTT broker with IP 192.168.8.138 on port 1883. If I check with 
a) MQTT explorer at 192.168.8.138 with port 1883 i get all the data/services from the venus OS but i cannot find any service /SBMS? what do i need to do in the MQTT broker of venus to receive this data from SBMS?
b) input MQTT in node the flow is connected to 192.168.8.138 port 1883 but the debug node SBMSRAW doesn't show any raw data, its empty.
i think i am missing a step here inbetween but what?

**Reply by diegotam855** [reply]
AHHHHH I know what's missing here! Now I remember that I struggled like you but it's been two years so.. 
**Any change you make in the SBMS0 settings needs a reboot to be effective.** Did you try to reboot the bms by disconnecting the balance cables?

I do not send any BMS data to the victron venus and vrm. I'm using it just inside node red to make some logics to control the victron devices. Simple PLC stuff. 
 
Here is the rest of the flow. It's specific for my situation though... 
[flows.json](https://github.com/Louisvdw/dbus-serialbattery/files/13891723/flows.json)

**Reply by CaptainRivet** [reply]
just rebooted it yesterday, done top balancing of the bank to 3,65V and currently running capacity test to see how much it degraded in last 3 years. have a 4P4S setup with 272AH lishen cells so nominal 1088AH and got it at 1170AH tested, lets see how that looks after 3years.
As soon as i got that sorted with the input data from SBMS i will 
a) get a nice dashboard started and also do bit stuff like control active balancer and other devices like waterheater based on voltages, the SOC steering in SBMS sucks 
b) dbus_mqTT-battery into venus OS.

**Reply by Tiry56** [reply]
Hi 
I can confirm that the SBMS0 needs a reboot to have the MQTT settings applied
on my side I was able to send data from SBMS0 to Venus OS MQTT broker, then have these read by the dbus-mqtt battery driver
but JSON structure sent by SBSM0 do not comply with the JSON structure required for dbus-mqtt battery driver unfortunatly
so the JSON structure of the SBSM0 needs to be adapted before it can be read by the dbus-mqtt battery driver

**Comment by CaptainRivet** [comment]
Have Electrodacus BMS (actually 2) installed on my boat and also have a cerbo GX.
Available for testing and can also ship an Electrodacus for development if needed.

**Comment by sin-tax** [comment]
Are there any updates on this integration? 

It would be great to tie both systems together!

**Reply by CaptainRivet** [reply]
Sadly no....will soon be the main  reason i ditch the Electrodacus for a jk or REC BMS as i need integration into cerbo for steering and also monitoring purposes via cerbo. 
The Electrodacus web interface is very unreliable...

**Reply by CaptainRivet** [reply]
Need the cell voltages and even for basic integration adding a smart shunt on negative cost another 130Euro and effort changing several things on negative side so it can be integrated. For 140 i get the jk BMS inkl. Integration into cerbo and simply connect the cell wires and main to terminals. Not riding a dead horse, a shame but without integration even the cheapest daly is ahead of Electrodacus

**Reply by mr-manuel** [reply]
You could use Node-RED and dbus-mqtt-battery from my repo to integrate this battery into Venus OS.

**Reply by sin-tax** [reply]
I kind of was afraid of that after i found a comment from the owner of this proj stating he was working away on it in 2021, and I didnt see it on the supported devices list. 

@Louisvdw  are there any technical challenges people in the community could help with, or is a lost cause all together due to limitations?

**Reply by mr-manuel** [reply]
The only limitation is time and effort to make this work. The more BMS this project support the more work it is to maintain all. We are doing this all in our spare time for free and spending lots of hours for this.

Adding a new BMS do take many hours of reading manuals, developing, testing, failing and fixing code.

Anyone of you is welcome to integrate the BMS by [opening a pull request](https://louisvdw.github.io/dbus-serialbattery/general/supported-bms#add-by-opening-a-pull-request).

**Reply by CaptainRivet** [reply]
@mr-manuel this only limitation is i have no glue about NoteRed and how this works. I can integrate a 600A constinous Lithium system in a boat but coding is simply not my world. If a have a dummy how to do that like with gui mods then i am capable of doing that. I assumed that in electrodacus community with a huge amount of victron devices this integration would be longtime done but i was wrong.

**Reply by CaptainRivet** [reply]
The reason is used ElectrodacusBMS is that you can use it for a hybrid starter and housebank where a Mosfet based BMS the starter and windlass surge destroyes the mosfets in the mosfet BMS. But well if i use 4 parallel then mosfet JK BMS works too.

**Reply by CaptainRivet** [reply]
Another question will this work with the new gui from Victron and i have an original cerbo GX, no raspi. Thought i do clean install and pay a bit more for the cerbo instead raspi with a lot adapters, also beacuse of the additional inputs like tank,relais.... the price difference is 50Euro in the end but a lot mods are just tested on raspi. And well if something goes wrong raspi format SD, put on new image on and start from scratch, not possible with cerbo.

**Reply by mr-manuel** [reply]
There is always the option to factory reset the GX device. No matter which one.

**Reply by CaptainRivet** [reply]
Thats sadly not the case. Eg had the white screen issue with the gui mod twice. Both time factory reset didn't work at all. Both times spend countless hours to get it fixed, without the help of Kevin i would have been screwed. wiping the SD card on raspi would have been much faster and possible without Kevin's help.

**Reply by mr-manuel** [reply]
This didn’t worked?

https://www.victronenergy.com/media/pg/Cerbo-S_GX/en/reset-to-factory-defaults-procedure.html

Then I would recommend you to setup SSH authentication with authentication keys, so you don‘t loose access after a failed update.

**Reply by CaptainRivet** [reply]
No didn’t work, it made it worse. What in the end worked was Kevins uninstaller (at a precise moment you needed to pull the memory stick so it coruptes someting that the installer then realized and reinstall new) and then via SSH with terminal access installing an older firmware on that then gave me access, then updating new. Problem with white screen you have no access to cerbo. Took me several days and countless hours, wiping SD card and start from scratch 1h with all adjustments. Meanwhile realize if you keep it stock cerbo but if you wanna modify raspi is the way to go.

**Reply by CaptainRivet** [reply]
And i am not allone, know min 20 boats that had the same issue and if you hadn't 2 way communication,  remote access monitoring and SSH on it would be bricked as you cannot access cerbo then to do the fixes. I made then a how to fix on cruisersforum. Later Kevin found the problem and fixed it with an new update.

**Comment by jpvardy** [comment]
+1 my vote for an integration!

**Comment by jpvardy** [comment]
Please upvote this discussion to help get this moving again:

https://github.com/mr-manuel/venus-os_dbus-serialbattery/discussions/91
